HW0/P4/P4.py

```diff
@@ -40,37 +49,124 @@
+
+
+    #sanity check
+    #print fixed_matrix
+
+    #extract coordinates
+    x_coords = fixed_matrix[:,0]
+    y_coords = fixed_matrix[:,1]
+    z_coords = fixed_matrix[:,2]
+
+    ax.plot(x_coords, y_coords, z_coords,
+             '.g', label='Observed trajectory')
 
     #####################
     # Part 3:
     # Use the initial conditions and propagation matrix for prediction
     #####################
 
-    # A = ?
-    # a = ?
-    # s = ?
+    A = np.matrix([[1, 0, 0, dt, 0, 0],
+                   [0, 1, 0, 0, dt, 0],
+                   [0, 0, 1, 0, 0, dt],
+                   [0, 0, 0, 1 - c*dt, 0, 0],
+                   [0, 0, 0, 0, 1 - c*dt, 0],
+                   [0, 0, 0, 0, 0, 1-c * dt]
+                    ])
+    a = np.matrix([0, 0, 0, 0, 0, g*dt]).transpose()
+    s = np.matrix([0, 0, 2, 15, 3.5, 4.0]).transpose()
+
 
     # Initial conditions for s0
+    prediction_matrix = np.zeros((6,121))
+    prediction_matrix[:,0] = s.reshape(6)
+
     # Compute the rest of sk using Eq (1)
+    for x in range (1, K):
+        prediction_matrix[:,x] = (A * prediction_matrix[:,x-1].reshape(6,1) + a).reshape(6)
+
+    #extract coordinates
+    x_coords = prediction_matrix[0,:]
+    y_coords = prediction_matrix[1,:]
+    z_coords = prediction_matrix[2,:]        
 
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '-k', label='Blind trajectory')
+    ax.plot(x_coords, y_coords, z_coords,
+             '-k', label='Blind trajectory')
 
     #####################
     # Part 4:
     # Use the Kalman filter for prediction
     #####################
 
-    # B = ?
-    # C = ?
+    B = np.diag([bx,by, bz, bvx, bvy, bvz])
+    C = np.matrix([[rx, 0, 0, 0, 0, 0],
+                  [0, ry, 0, 0, 0, 0],
+                  [0, 0, rz, 0, 0, 0]])
+
+
+    observation = np.loadtxt("P4_measurements.txt", delimiter=',')
+
+    #convert observation array into a matrix
+    ob_matrix = np.matrix(observation)
 
     # Initial conditions for s0 and Sigma0
+
+
+    s0 = np.matrix([0, 0, 2, 15, 3.5, 4.0]).transpose()
+    sigma = 0.01 * np.identity(6)
+
+
+    filtered_matrix = np.zeros((6,121))
+    filtered_matrix[:,0] = s0.reshape(6)
+
     # Compute the rest of sk using Eqs (2), (3), (4), and (5)
 
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '-r', label='Filtered trajectory')
+    for x in range(1,K):
+        #equation 2: predictS
+        s_tilde = (A * filtered_matrix[:,x-1].reshape(6,1) + a).reshape(6).transpose()
+
+        #equation 3: predictSig
+        sigma_tilde = inv(A * sigma * A.transpose() + B*B.transpose())
+
+        #equation 4: update sig
+        sig_updated = inv(sigma_tilde + C.transpose() * C)
+
+        #equation 5: updateS
+        s_updated = sig_updated * (sigma_tilde * s_tilde + C.transpose() *  ob_matrix[x,:].transpose())
+
+        filtered_matrix[:,x] = s_updated.reshape(6)     
+
+        #save sigma for next iteration
+        sigma = sig_updated   
+
```

Comment:
It would have been easier to just call recalculate sigma in each for loop instead of calling it sig_updated and then equating sigma to it again. I understand the idea of updating the value at each loop, and hence using a sig_updated variable, but it would be more efficient to directly assign the sigma updated value.
You could directly have: sigma = inv(sigma_tilde + C.transpose() * C)

Instead of : sig_updated = inv(sigma_tilde + C.transpose() * C)
sigma=sig_updated